### PR TITLE
BUG: minimize should work with a scalar closes #15228

### LINKED ISSUE #15228: BUG: bounded L-BFGS-B doesn't work with a scalar.
### Describe your issue.

The  following example worked with 1.7.*, but not with version 1.8. @tylerjereddy, this needs to be fixed before the final release of 1.8.0. I have a PR incoming.

### Reproducing Code Example

```python
import numpy as np
from scipy.optimize import minimize, rosen

def f(x):
    return np.sum(x**2)
    
x0 = 17
res = minimize(f, x0, bounds=[(-100, 100)], method='L-BFGS-B')
```

### Error message

```shell
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
/var/folders/m8/gwtcncws12jf60xw5n6knwnw0000gn/T/ipykernel_92447/1079589486.py in <module>
      6 
      7 x0 = 17
----> 8 res = minimize(f, x0, bounds=[(-100, 100)], method='L-BFGS-B')
      9 print(res)

~/Documents/Andy/programming/scipy/scipy/optimize/_minimize.py in minimize(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)
    664                                                        x_fixed, min_dim=2,
    665                                                        remove=1)
--> 666         bounds = standardize_bounds(bounds, x0, meth)
    667 
    668     if meth == 'nelder-mead':

~/Documents/Andy/programming/scipy/scipy/optimize/_minimize.py in standardize_bounds(bounds, x0, meth)
    927     elif meth in ('l-bfgs-b', 'tnc', 'slsqp', 'old'):
    928         if isinstance(bounds, Bounds):
--> 929             bounds = new_bounds_to_old(bounds.lb, bounds.ub, x0.shape[0])
    930     return bounds
    931 

IndexError: tuple index out of range
```

### SciPy/NumPy/Python version information

1.8.dev

### PR DESCRIPTION
**Pull request by andyfaff**
@tylerjereddy, this should be a backport to 1.8.0 before it gets released.

### CODE REVIEW

**Review by mdhaber — APPROVED** [review]
I think this is all it takes.

One concern I considered is whether user code that might have been passed a 0D array before this change might fail with the new 1D array. I don't think that is credible for two reasons:
* User code was never passed a 0D array. I checked in 1.7, and even when `x0` passed into `minimize` is a scalar, a 1D array is passed into `fun` for all of these optimization methods. I'm not sure where it gets converted to 1D for all of them, but for `l-bfgs-b`, for instance, it happens in `_minimize_lbfgsb`: `x0 = asarray(x0).ravel()`
* I can't think of normal operations that would work on a 0D array and fail on a 1D array. (The other way around doesn't always work, of course.)

The other concern was whether `res.x` might have been a 0D array in the past, and now it will be 1D. It was 1D for all these methods in SciPy 1.7.

So LGTM.

Anything to add, @andyfaff? Any concerns that aren't mentioned here that we should consider before merging?

File: scipy/optimize/tests/test_optimize.py
Comment (mdhaber) [inline scipy/optimize/tests/test_optimize.py:1428]:
```suggestion
    'method',
    ['l-bfgs-b', 'tnc', 'Powell', 'Nelder-Mead']
```

**Comment by andyfaff** [comment]
> User code was never passed a 0D array. I checked in 1.7, and even when x0 passed into minimize is a scalar, a 1D array is passed into fun for all of these optimization methods. I'm not sure where it gets converted to 1D for all of them, but for l-bfgs-b, for instance, it happens in _minimize_lbfgsb: x0 = asarray(x0).ravel()
I can't think of normal operations that would work on a 0D array and fail on a 1D array. (The other way around doesn't always work, of course.)

I shared some of the same concerns, but I think it's a reasonable assumption that `minimize` should always return a `res.x` that is an array, because it's minimising `(n,)` variables. If people want to minimize a scalar they should be using `minimize_scalar`. I'm happy to see it merged, I added your lint suggestion.

**Comment by mdhaber** [comment]
I'll wait for CI to finish and then merge.

**Review by tylerjereddy — COMMENTED** [review]
Alright, patch diff coverage is 100 % and CI failures are just the usual timeouts.

The regression test seems to match the original issue report to me.

The discussion is a bit complex, but there is an approval from Matt and an intention to merge after CI ran, so I'll go ahead and put this in. Thanks @andyfaff @mdhaber

**Comment by tylerjereddy** [comment]
squash merge to facilitate backport

**Comment by mdhaber** [comment]
Thanks, Tyler! Got distracted with CI timeouts.
Thanks for the fix @andyfaff!

**Pull request merged**
